ingest/annotations.py

```diff
@@ -83,17 +85,26 @@ def merge_df(self, first_df, second_df):
         """ Does an inner join on a dataframe """
         self.file = pd.merge(second_df, first_df, on=[("NAME", "TYPE")])
 
-    def preprocess(self):
-        """Ensures that:
+    def preprocess(self, is_metadata_convention=False):
+        """ Prepares file for ingest
+        Ensures that:
             - 'NAME' in first header row is capitalized
             - 'TYPE' in second header row is capitalized
+            - Numeric and group annotation values are coerced properly
         """
 
         # Uppercase NAME and TYPE
         self.headers[0] = self.headers[0].upper()
         self.annot_types[0] = self.annot_types[0].upper()
         if self.validate_unique_header():
             self.create_data_frame()
+            self.coerce_empty_numeric_values()
+            # Metadata convention can contain arrays that have numeric or string values.
+            # Therefore dtypes for numeric annotations are skipped.
```

Comment:
Great question! 
The only thing ingest knows is the annotation name and type and if the file is conventional. When a column has array-based values, the annotation type refers to the values in the array. For example, an annotation with arrays and an annotation type numeric says that the array has numeric values. For conventional files, we can't coerce numeric columns because they may contain arrays, and arrays cannot be turned into floats.
This class does not touch any of the convention validation. You're allowed to have additional columns outside of the required columns for convention files. validate_metadat.py takes care of all the magic that happens there. Also, it takes care of reading and interpreting array-based values.
For cell metadata convention does not affect group annotations. So we force the values to strings in [create_dataframe() ](https://github.com/broadinstitute/scp-ingest-pipeline/blob/52e56c70ea745dcb7ea7f30877e6ed35165237ea/ingest/annotations.py#L190) by calling [coerce_group_values](https://github.com/broadinstitute/scp-ingest-pipeline/blob/52e56c70ea745dcb7ea7f30877e6ed35165237ea/ingest/annotations.py#L133)